### DISCUSSION TITLE
Why preset-react uses unexisted React.createElement when runtime is classic?

### DISCUSSION THREAD

**Discussion by krutoo** [babel/babel#16027] [Q&A] [answered]
Hi, I use this config:
```js
module.exports = {
  presets: [
    [
      '@babel/preset-env',
      {
        modules: false,
        targets: {
          node: true,
        },
      },
    ],
    [
      '@babel/preset-react',
      {
        runtime: 'classic', // for both of react 17 and 18 
      },
    ],
    '@babel/preset-typescript',
  ],
};
```

And in result files **there is no import of reactElement function from react package**

Example of unexpected behavior: [babel try it out](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=KYDwDg9gTgLgBAMwK4DsDGMCWEVwLICeAwhALaQrArwAUAlHAN4BQAkFMDElLgDwAmmAG4A-ABLAANpIgBCXgHpBo5gF8gA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Creact%2Cstage-3&prettier=false&targets=&version=7.23.1&externalPlugins=&assumptions=%7B%7D)

```js
export function MyComponent() {
  return /*#__PURE__*/React.createElement("div", null, "Hello!");
}
```

How can i fix this and add valid imports to result files?

From:
```jsx
export function MyComponent () {
  return <div>Hello!</div>
}
```

To:
```js
import { createElement } from 'react';

export function MyComponent() {
  return createElement("div", null, "Hello!");
}
```

**Comment by nicolo-ribaudo** [comment]
Because classic React compilation (in every JSX compiler, not just Babel) expected you to manually add `import React from "react"` in your file whenever you want to use JSX.

This has been "fixed" with `runtime: "automatic"`, that automatically adds imports for you.

**Reply by krutoo** [reply]
@nicolo-ribaudo but i want to use automatic runtime in source code and classic in result code

is it possible to do with babel reset react?

**Comment by krutoo** [comment, marked as answer]
I use custom transformation after build that looks like:

```js
import fs from 'node:fs/promises';
import glob from 'fast-glob';

async function prependFile(path, content) {
  const data = await fs.readFile(path);
  const handle = await fs.open(path, 'w+');
  const insert = Buffer.from(content);

  await handle.write(insert, 0, insert.length, 0);
  await handle.write(data, 0, data.length, insert.length);
  await handle.close();
}

const react = `import * as React from 'react';\n`;

glob('./dist/esm/**/*.js')
  .then(items =>
    Promise.all(
      items.map(item => fs.readFile(item, 'utf-8').then(content => ({ filename: item, content }))),
    ),
  )
  .then(files => files.filter(item => item.content.includes('React.createElement')))
  .then(files => Promise.all(files.map(file => prependFile(file.filename, react))));

```
